jax/_src/api_util.py

```diff
@@ -495,6 +496,19 @@ def infer_argnums_and_argnames(
   return argnums, argnames
 
 
+def _is_static(annotation: Any) -> bool:
+  return getattr(annotation, "__metadata__", None) == (StaticTag,)
+
+
+def static_argnames_from_annotations(fun: Callable[..., Any]) -> tuple[str, ...]:
+  try:
+    hints = get_type_hints(fun, include_extras=True)
+  except (TypeError, ValueError, NameError):
```

Comment:
Not sure if we should suppress `NameError`. IIUC this can only happen if a forward reference is local or if a type annotation references an undefined name.

Reply:
I got a `NameError` when running the JAX test suite and this was my fix.

Reply:
I'm just concerned that we would ignore `Static` annotations if we handle `NameError` unconditionally. Would it be a hack to check if `__annotations__` doesn't contain `Static[` as a substring on `NameError` and re-raise if that's the case?

Reply:
Sure, I'll give that a try.

Reply:
Just pushed a commit that tries to do smarter error handling, with detailed comments. It's a bit messy to be honest... let me know what you think.